### ISSUE TITLE
[BUG] TopKRouter expert_bias get wrong dtype when updated

### ISSUE THREAD

**Issue by iansheng** [NVIDIA/Megatron-LM#1482]
**Describe the bug**
Although the buffer `expert_bias` in `TopkRounter` was registered with float32, it is observed to be in bfloat16 format during the forward pass and `finalize_model_grads` process when i pretrain in bf16.

The expert_bias definition:

https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/transformer/moe/router.py#L115-L123

**To Reproduce**
Enable `--moe-router-enable-expert-bias` and `--moe-router-score-function sigmoid` ,  and pretrain with `--bf16`.

> MOE_ARGS="
>     --moe-grouped-gemm \
>     --moe-token-dispatcher-type alltoall \
>     --moe-shared-expert-intermediate-size 2048 \
>     --num-experts 16 \
>     --moe-router-topk 8 \
>     --moe-ffn-hidden-size 2048 \
>     --moe-router-load-balancing-type seq_aux_loss \
>     --moe-router-pre-softmax \
>     --moe-router-topk-scaling-factor 2.5 \
>     --moe-aux-loss-coeff 1e-2 \
>     --moe-router-enable-expert-bias \
>     --moe-router-score-function sigmoid \
>     --moe-router-dtype fp32

**Expected behavior**

Expert bias should always use the float32 format, especially it might be updated at a very small rate.

**Stack trace/logs**

![Image](https://github.com/user-attachments/assets/7ddd7115-3ec8-4ca6-a7ff-d43f801816f1)

**Environment (please complete the following information):**
 - Megatron-LM main
 - PyTorch 2.1.0
 - CUDA 12.6
 - NCCL 2.22.3

**Proposed fix**
If you have a proposal for how to fix the issue state it here or link to a PR.

**Additional context**

**Comment by iansheng** [comment]
Hi, @yanring @Victarry can you check this out when you have a moment?

**Comment by Victarry** [comment]
Thanks for reporting. I will take a look.

**Comment by jiurizz** [comment]
`module.bfloat16()` in Float16Module converts the expert_bias to bf16

my solution is override the _apply in TopKRouter class:

```
  def recover_fp32(self):
      self.expert_bias = self.expert_bias.to(torch.float32)
      self.local_tokens_per_expert = self.local_tokens_per_expert.to(torch.float32)

  def _apply(self, fn, recurse=True):
      super()._apply(fn, recurse)
      self.recover_fp32()
      return self
```

**Comment by Victarry** [comment]
@iansheng @jiurizz  Thanks for your helpful feedbacks!
I did a small experiment and found when the expert_bias > 0.5, the update will have no effects due to the rounding error of bf16.
```
>>> import torch
>>> torch.tensor(0.5, dtype=torch.bfloat16) + 1e-3
tensor(0.5000, dtype=torch.bfloat16)
```
I will submit a PR to resolve this issue.

BTW, do you have any training loss curve that can be shared to show the influence of expert_bias dtypes? Thanks a lot~

**Comment by jiurizz** [comment]
I discovered this bug when I updated this feature, so there is no data to provide

**Comment by iansheng** [comment]
https://github.com/NVIDIA/Megatron-LM/commit/8dd03a39ccc67a93c186cfe16fa782ac7144480d
Noted the submission. Thanks!
